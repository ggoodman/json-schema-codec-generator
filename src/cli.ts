import Yargs from 'yargs';
import * as Path from 'path';
import { promises as Fs } from 'fs';
import type { Schemas } from './index';
import fastGlob from 'fast-glob';

Yargs.help()
  .strict()
  .showHelpOnFail(true)
  .command(
    '$0 [path]',
    false,
    (args) => {
      return args.options({
        input: {
          alias: 'i',
          string: true,
          description: 'Path to a directory containing your json-schema files.',
          demandOption: true,
        },
        output: {
          alias: 'o',
          string: true,
          description:
            'Path to a directory where the generated codecs will be written. If not specified, the generated ',
        },
      });
    },
    async ({ input, output }) => {
      const inputPath = Path.resolve(process.cwd(), input);

      const matches = await fastGlob('**/*.json', {
        cwd: inputPath,
        stats: false,
        ignore: input.includes('node_modules') ? [] : ['node_modules/**/*.json'],
        absolute: false,
      });

      if (!matches.length) {
        console.error(`‚ùå No .json files found in ./${Path.relative(process.cwd(), inputPath)}.`);
        process.exit(1);
      }

      const schemas: Schemas = {};

      for (const schemaPathRel of matches) {
        const schemaPath = Path.join(inputPath, schemaPathRel);
        const schemaUrl = `file:///${schemaPathRel}`;
        const schemaData = await Fs.readFile(schemaPath, 'utf8');
        const schema = JSON.parse(schemaData);

        schemas[schemaUrl] = schema;
      }

      const { generateCodecCode } = await import('./index');
      const generated = await generateCodecCode(schemas);

      if (typeof output === 'string') {
        const outputPath = Path.resolve(process.cwd(), output);
        await Promise.all([
          Fs.writeFile(Path.join(outputPath, 'index.js'), generated.javaScript),
          Fs.writeFile(Path.join(outputPath, 'index.d.ts'), generated.typeDefinitions),
        ]);
      } else {
        console.log(JSON.stringify(generated, null, 2));
      }
    }
  ).argv;
