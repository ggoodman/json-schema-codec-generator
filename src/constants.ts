export const VIRTUAL_ROOT = '/virtual'
