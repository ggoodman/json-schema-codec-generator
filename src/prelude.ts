export const prelude = `
export interface Codec<T> {
  /**
   * Validate that an unknown value matches the underlying schema, returning a typed
   * object.
   */
  validate(json: unknown): T;

  /**
   * Return the input value unmodified while asserting the input is correctly typed.
   * 
   * The identify function is useful to use in untyped code to facilitate
   * constructing valid values while benefitting from the type-checking
   * features of modern code editors. This function may also be useful to act as a
   * compile-time type guard that imposes no runtime validation cost, since the input
   * is simply returned unmodified.
   */
  identity(value: T): T;

  /**
   * Validate that an unknown value matches the underlying schema, and act as a boolean
   * type-guard.
   */
  is(json: unknown): json is T;
}

interface ValidatorError {
  keywordLocation: string;
  instanceLocation: string;
}

interface Validator<T> {
  (json: unknown): json is T;
  errors: ValidatorError[] | null;
}

class CodecImpl<T> implements Codec<T> {
  constructor(
    public readonly name: string,
    private readonly validator: Validator<T>
  ) {}

  validate(value: unknown): T {
    if (!this.validator(value)) {
      throw this.validationError(value, this.validator.errors);
    }

    return value;
  }

  identity(value: T): T {
    return value;
  }

  is(json: unknown): json is T {
    try {
      this.validate(json);
      return true;
    } catch {
      return false;
    }
  }

  private validationError(value: unknown, errors: ValidatorError[] | null) {
    // Defensive code to check that schemasafe produced the errors as expected
    if (!Array.isArray(errors) || !errors.length) {
      return new Error(\`Value did not match the schema \${this.name}\`);
    }

    return new ValidationError(this.name, this.schema, value, errors);
  }
}

export class ValidationError extends Error {
  static isValidationError(err: unknown): err is ValidationError {
    return err instanceof this;
  }

  public readonly schema: unknown;
  public readonly value: unknown;
  public readonly validatorErrors: ReadonlyArray<ValidatorError>;

  constructor(
    schemaName: string,
    schema: unknown,
    value: unknown,
    validatorErrors: ValidatorError[]
  ) {
    const seenValuePaths = new Set<string>();
    const errorStrings: string[] = [];

    nextError: for (const validatorError of validatorErrors) {
      const valuePath = locationToPath(validatorError.instanceLocation);
      const [foundPath, valueSlice] = getAtClosestPath(value, valuePath);
      const foundPathString = \`#/\${foundPath.join('/')}\`;

      if (seenValuePaths.has(foundPathString)) {
        continue;
      }

      for (const seenValuePath of seenValuePaths) {
        if (foundPathString.startsWith(\`\${seenValuePath}/\`)) {
          continue nextError;
        }
      }

      seenValuePaths.add(foundPathString);

      const shape = valueToShapeString(valueSlice);

      errorStrings.push(
        \`  The value at \${JSON.stringify(
          foundPathString
        )} with shape \${shape} did not match the schema defined at \${JSON.stringify(
          validatorError.keywordLocation
        )}\`
      );
    }

    super(
      \`Validation for the schema \${JSON.stringify(
        schemaName
      )} failed with the following errors:\n\${errorStrings.join('\n')}\`.replace(/"/g, "'")
    );

    this.schema = schema;
    this.value = value;
    this.validatorErrors = validatorErrors;
  }
}

/**
* Convert a json-path reference to an array of string segments
*
* @param location json-path such as \`#/path/to/value\`
*/
function locationToPath(location: string): string[] {
  const unprefixedLocation = location.startsWith('#/') ? location.slice(2) : location;

  return unprefixedLocation.split('/');
}

function getAtClosestPath(value: unknown, path: string[]): [string[], unknown] {
  const foundPath: string[] = [];
  let lastValue = value;

  for (let segment of path) {
    const nextValue = (lastValue as any)[segment] as unknown;

    if (typeof nextValue === 'undefined' || nextValue === null) {
      break;
    }

    foundPath.push(segment);
    lastValue = nextValue;
  }

  return [foundPath, lastValue];
}

function valueToShapeString(value: unknown) {
  return JSON.stringify(value, valueToShapeReplacer);
}

function valueToShapeReplacer(_key: string, value: unknown) {
  return typeof value === 'object' && value ? value : typeof value;
}
`;
