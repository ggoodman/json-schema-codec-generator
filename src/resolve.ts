import resolveSpec, { AsyncOpts } from 'resolve';

export function resolveAsync(spec: string, options: AsyncOpts) {
  return new Promise<string | undefined>((resolve, reject) =>
    resolveSpec(spec, options, (err, resolved) => {
      if (err && (err as any).code !== 'MODULE_NOT_FOUND') {
        return reject(err);
      }

      return resolve(resolved);
    })
  );
}
