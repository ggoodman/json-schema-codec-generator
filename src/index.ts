import RollupPluginCommonJs from '@rollup/plugin-commonjs';
import RollupPluginTs from '@wessberg/rollup-plugin-ts';
import Ajv, { Options } from 'ajv';
import standaloneCode from 'ajv/dist/standalone';
import { JSONSchema, Parser } from 'json-schema-to-dts';
import * as Path from 'path';
import { format } from 'prettier';
import { OutputAsset, OutputChunk, rollup } from 'rollup';
import { VIRTUAL_ROOT } from './constants';
import { staticFiles } from './embedded';

export { CodecImpl } from './stub/codec';
export type { Codec } from './stub/codec';
export { ValidationError } from './stub/validator';

const indexPrelude =
  `
import { CodecImpl } from './codec';
import { Codec } from './codec';
import * as Types from './types';
import * as Validators from './validators';
export { Codec, Types };
`.trim() + '\n';
const validationPrelude =
  `
import * as Types from './types';
import { ValidateFunction } from './validator';  
`.trim() + '\n';

export interface SchemaEntry {
  uri: string;
  schema: Exclude<JSONSchema, boolean>;
  preferredName?: string;
}
export type Schemas = [SchemaEntry];

export interface GenerateCodecCodeOptions
  extends Omit<Options, 'allErrors' | 'code' | 'inlineRefs'> {}

export async function generateCodecCode(schemas: Schemas, options: GenerateCodecCodeOptions = {}) {
  const parser = new Parser();
  const ajv = new Ajv({
    verbose: true,
    ...options,
    allErrors: true,
    code: {
      es5: false, // use es6
      lines: true,
      optimize: false, // we'll let rollup do this
      source: true,
    },
    inlineRefs: false,
  });

  const exportedNameToSchema: Record<string, JSONSchema> = {};
  const uriToExportedName: Record<string, string> = {};
  /**
   * A collection of exported names used to add safety to some risky
   * regex-based code rewriting of `exports.<symbol> = ` to `export const <symbol> = `.
   */
  const exportedNames = new Set<string>();
  const validationFunctionDefinitions: string[] = [];
  const codecDefinitions: string[] = [];

  for (const { schema, uri, preferredName } of schemas) {
    const schemaWithId: JSONSchema = { $id: uri, ...schema };
    const name = parser.addSchema(uri, schemaWithId, { preferredName });

    if (exportedNames.has(name)) {
      throw new Error(
        `Invariant violation: The name ${JSON.stringify(name)} was expored more than once`
      );
    }

    exportedNames.add(name);
    uriToExportedName[uri] = name;
    exportedNameToSchema[name] = schemaWithId;

    ajv.addSchema(schemaWithId, name);

    validationFunctionDefinitions.push(`export const ${name}: ValidateFunction<Types.${name}>;`);
    codecDefinitions.push(
      `${name}: new CodecImpl<Types.${name}>(${JSON.stringify(name)}, ${JSON.stringify(
        uri
      )}, Validators.${name}) as Codec<Types.${name}>`
    );
  }

  const { diagnostics, text: schemaTypeDefs } = parser.compile({
    topLevel: {
      hasDeclareKeyword: false,
      isExported: true,
    },
  });

  if (diagnostics.length) {
    throw new Error(
      `Produced diagnostics while generating type definitions for schemas: ${diagnostics
        .map((diagnostic) => `${diagnostic.message}`)
        .join('\n')}`
    );
  }

  const standaloneValidationCode = standaloneCode(ajv);

  // There appears to be a bug in AJV code generation that writes the same function
  // more than once to the resulting validation code when there are $refs between
  // multiple schemas. This isn't strictly wrong since JavaScript allows redefinition
  // of functions but causes babel / TypeScript to barf.
  // We use this to track validation function names that are observed so that if
  // we find one that has already been seen, we can replace it with a dummy identifier.
  const seenFunctionNames = new Set<string>();
  let nextDummyFunctionSuffix = 0;

  // The generated standalone validation code is in CommonJS. In order to
  // make a slimmer build, we're going to rewrite this to ESM. We replace
  // CommonJS export statements corresponding to entries in `exportedNames`
  // with ESM constant exports.
  const validationCode = standaloneValidationCode
    .replace(/^(?:module\.\s*)?exports.(\w+)\s*=/gm, (match, exportName) => {
      if (!exportedNames.has(exportName)) {
        throw new Error(
          `Invariant violation: The generated validation code contains the following CommonJS export ${JSON.stringify(
            match
          )} but there is no known exported type ${JSON.stringify(exportName)}`
        );
      }

      return `export const ${exportName} =`;
    })
    // We also need to rewrite some CommonJS imports added by AJV
    .replace(/const\s+(\S+)\s*=\s*require\(([^)]+)\).default/gm, (match, importName, spec) => {
      return `import ${importName} from ${spec}`;
    })
    .replace(/^function validate\d+/gm, (match) => {
      if (seenFunctionNames.has(match)) {
        // These 'dummy' functions should be eliminated by Rollup's tree shaking
        return `${match}ShakeMePlease${nextDummyFunctionSuffix++}`;
      }

      seenFunctionNames.add(match);

      return match;
    });

  const compilerOptions = {
    allowJs: true,
    allowSyntheticDefaultImports: true,
    alwaysStrict: true,
    baseUrl: 'src',
    checkJs: false,
    declaration: true,
    declarationMap: false,
    downlevelIteration: false,
    esModuleInterop: true,
    // isolatedModules: true,
    importHelpers: true,
    lib: ['es5'],
    skipDefaultLibCheck: true,
    skipLibCheck: true,
    module: 'commonjs',
    moduleResolution: 'node',
    outDir: 'dist',
    rootDir: 'src',
    target: 'es2017',
  };
  const tsConfig = {
    include: ['src'],
    compilerOptions,
  };

  const vfs = { ...staticFiles };

  vfs[Path.join(VIRTUAL_ROOT, 'src/index.ts')] =
    `
    ${indexPrelude}

    export const Codecs = {
    ${codecDefinitions.join(',\n')}
    } as const;
  `.trim() + '\n';

  vfs[Path.join(VIRTUAL_ROOT, 'src/types.ts')] =
    `
    ${schemaTypeDefs}
  `.trim() + '\n';

  vfs[Path.join(VIRTUAL_ROOT, 'src/validators.js')] =
    `
    ${validationCode}
  `.trim() + '\n';

  vfs[Path.join(VIRTUAL_ROOT, 'src/validators.d.ts')] =
    `
    ${validationPrelude}
    ${validationFunctionDefinitions.join('\n')}
  `.trim() + '\n';

  vfs[Path.join(VIRTUAL_ROOT, 'tsconfig.json')] = JSON.stringify(tsConfig);

  const build = await rollup({
    input: Path.resolve(VIRTUAL_ROOT, 'src/index.ts'),
    treeshake: {
      moduleSideEffects: (id) => {
        // console.debug('moduleHasSideEffects(%s, %s)', id, isExternal);

        // Make sure unreferenced modules can be tree-shaken by forcing
        // them to be considered side-effect-free.
        return false;
      },
    },
    plugins: [
      virtualFileSystemPlugin(vfs),
      RollupPluginTs({
        cwd: VIRTUAL_ROOT,
        transpiler: 'babel',
        browserslist: 'node 10',
        tsconfig: compilerOptions,
        // transpileOnly: true,
        fileSystem: createVirtualFilesystem(vfs),
      }),
      RollupPluginCommonJs({
        transformMixedEsModules: true,
      }),
      {
        name: 'prettier',
        renderChunk: async (code, chunk) => {
          return {
            code: format(code, {
              filepath: Path.join(process.cwd(), chunk.fileName),
            }),
          };
        },
      },
    ],
    onwarn: (warning) => {
      console.warn(warning.message);
    },
  });

  const output = await build.generate({
    dir: Path.resolve(VIRTUAL_ROOT, 'dist'),
    format: 'cjs',
  });

  const javaScriptChunk = output.output.find(
    (chunk) => chunk.type === 'chunk' && chunk.fileName === 'index.js'
  ) as OutputChunk | undefined;
  if (!javaScriptChunk) {
    throw new Error(`Invariant violation: The build failed to produce a JavaScript bundle.`);
  }

  const typeDefinitionChunk = output.output.find(
    (chunk) => chunk.type === 'asset' && chunk.fileName === 'index.d.ts'
  ) as OutputAsset | undefined;
  if (!typeDefinitionChunk) {
    throw new Error(`Invariant violation: The build failed to produce a JavaScript bundle.`);
  }

  return {
    javaScript: javaScriptChunk.code,
    typeDefinitions: typeDefinitionChunk.source,
    schamaPathsToCodecNames: uriToExportedName,
  };
}

function virtualFileSystemPlugin(vfs: { [key: string]: string }): import('rollup').Plugin {
  return {
    name: 'vfs',
    resolveId: async (source, importer) => {
      const candidateExt = ['', '.js', '.ts'];
      const [sourceBase, suffix = ''] = source.split('?', 2);

      let resolved = importer ? Path.resolve(Path.dirname(importer), sourceBase) : sourceBase;

      if (!sourceBase.startsWith('/') && !sourceBase.startsWith('.')) {
        resolved = `/virtual/src/node_modules/${sourceBase}`;
        // Solve for the case where node will consider a directory with an index also
        candidateExt.push(...candidateExt.map((ext) => `/index${ext}`));
      }

      for (const ext of candidateExt) {
        const candidate = `${resolved}${ext}`;

        if (candidate in vfs) {
          // console.debug('vfs.resolveId(%s, %s): %s', sourceBase, importer, candidate, '✅');
          return `${candidate}${suffix}`;
        }
        // console.debug('vfs.resolveId(%s, %s): %s', sourceBase, importer, candidate, '❌');
      }

      // console.debug('vfs.resolveId(%s, %s): %s', sourceBase, importer, undefined);
    },
    load: async (source) => {
      // console.debug('vfs.load(%s)', source);
      return vfs[source];
    },
  };
}

function createVirtualFilesystem(
  vfs: Record<string, string>
): import('@wessberg/rollup-plugin-ts').TypescriptPluginOptions['fileSystem'] {
  const virtualFs: import('@wessberg/rollup-plugin-ts').TypescriptPluginOptions['fileSystem'] = {
    directoryExists: (dirName) =>
      Object.keys(vfs).some((pathName) => pathName.startsWith(`${dirName}/`)),
    ensureDirectory: (dirName) => dirName,
    fileExists: (fileName) => fileName in vfs,
    getDirectories: (dirName) => {
      const dirNames = new Set<string>();
      const prefix = `${dirName}/`;

      for (const fileName in vfs) {
        if (fileName.startsWith(prefix)) {
          const rest = fileName.slice(prefix.length).split('/');

          if (rest.length > 1) {
            dirNames.add(rest[0]);
          }
        }
      }

      return [...dirNames];
    },
    newLine: '\n',
    readDirectory: (dirName, extensions, excludes, includes, depth = 20) => {
      const prefix = `${dirName}/`;
      const readResults = new Set<string>();

      for (const fileName in vfs) {
        if (fileName.startsWith(prefix)) {
          const rest = fileName.slice(prefix.length).split('/');

          readResults.add(`${prefix}${rest.slice(0, depth).join('/')}`);
        }
      }

      return [...readResults];
    },
    readFile: (fileName) => vfs[fileName],
    realpath: (fileName) => fileName,
    useCaseSensitiveFileNames: true,
    writeFile: (fileName, data) => void (vfs[fileName] = data),
  };

  return new Proxy(virtualFs, {
    get: (target, p) => {
      const found = Reflect.get(target, p);

      return typeof found === 'function'
        ? (...args: any[]) => {
            const res = found(...args);
            // console.debug(
            //   p,
            //   ...args,
            //   '=>',
            //   typeof res === 'string' || Buffer.isBuffer(res) ? res.toString().slice(0, 30) : res
            // );

            return res;
          }
        : found;
    },
  });
}

// async function main() {
//   return await withCleanup(async (defer) => {
//     const tmpDir = await Fs.mkdtemp('codec-', { encoding: 'utf-8' });
//     defer(() => Fs.rmdir(tmpDir, { recursive: true, maxRetries: 2 }));

//     for await (const entry of Fs.readdir())

//   });
// }
