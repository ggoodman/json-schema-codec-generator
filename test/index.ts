/// <reference types="jest" />

import { generateCodecCode } from '../src';

describe('Codec generation', () => {
  it('will produce the expected code', async () => {
    const { javaScript, schamaPathsToCodecNames, typeDefinitions } = await generateCodecCode([
      {
        schema: {
          title: 'A User Object',
          description: 'A user is a known visitor.',
          type: 'object',
          properties: {
            id: {
              type: 'string',
            },
            name: {
              type: 'string',
            },
          },
          required: ['id', 'name'],
        },
        uri: 'file:///User.json',
        preferredName: 'User',
      },
      {
        schema: {
          title: 'A Blog Post',
          description: 'A blog post represents an article associated with an author',
          type: 'object',
          properties: {
            id: {
              type: 'string',
            },
            title: {
              type: 'string',
            },
            content: {
              type: 'string',
            },
            author: {
              $ref: 'file:///User.json',
            },
          },
          required: ['id', 'title', 'content', 'user'],
        },
        uri: 'file:///BlogPost.json',
        preferredName: 'BlogPost',
      },
    ]);

    expect(javaScript).toMatchSnapshot();
    expect(schamaPathsToCodecNames).toMatchSnapshot();
    expect(typeDefinitions).toMatchSnapshot();
  });
});
