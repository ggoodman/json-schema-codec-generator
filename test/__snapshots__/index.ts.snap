// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Codec generation will produce the expected code 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", { value: true });

class ValidationError extends Error {
  constructor(schemaName, value, validatorErrors) {
    const errorStrings = validatorErrors.map((err) => {
      return \`  \${err.message} at \${
        err.dataPath || \\"#\\"
      }, got \${valueToShapeString(err.data)}\`;
    });
    super(
      \`Validation for the schema \${JSON.stringify(
        schemaName
      )} failed with the following errors:\\\\n\${errorStrings.join(\\"\\\\n\\")}\`
    );
    this.value = value;
    this.validatorErrors = validatorErrors;
  }

  static isValidationError(err) {
    return err instanceof this;
  }
}

function valueToShapeString(value) {
  return JSON.stringify(value, valueToShapeReplacer);
}

function valueToShapeReplacer(_key, value) {
  return typeof value === \\"object\\" && value ? value : typeof value;
}

class CodecImpl {
  constructor(name, uri, validateFn) {
    this.name = name;
    this.uri = uri;
    this.validateFn = validateFn;
  }
  /**
   * Identify function returning the given argument as a value matching the schema.
   *
   * This can be useful to use in non-TypeScript code to construct a valid object while
   * benefitting from suggestions from a TypeScript language service.
   */

  identity(obj) {
    return obj;
  }
  /**
   * Check if a value matches the schema.
   */

  is(obj) {
    return this.validateFn(obj);
  }
  /**
   * Validate that a value matches the schema and throws if not.
   */

  validate(obj) {
    if (!this.validateFn(obj)) {
      throw new ValidationError(this.name, obj, this.validateFn.errors || []);
    }

    return obj;
  }
}

var types = /*#__PURE__*/ Object.freeze({
  __proto__: null,
});

const User = validate21;
const schema6 = {
  $id: \\"file:///User.json\\",
  title: \\"A User Object\\",
  description: \\"A user is a known visitor.\\",
  type: \\"object\\",
  properties: {
    id: {
      type: \\"string\\",
    },
    name: {
      type: \\"string\\",
    },
  },
  required: [\\"id\\", \\"name\\"],
};

function validate21(
  data,
  { dataPath = \\"\\", parentData, parentDataProperty, rootData = data } = {}
) {
  let vErrors = null;
  let errors = 0;

  if (data && typeof data == \\"object\\" && !Array.isArray(data)) {
    if (data.id === undefined) {
      const err0 = {
        keyword: \\"required\\",
        dataPath,
        schemaPath: \\"#/required\\",
        params: {
          missingProperty: \\"id\\",
        },
        message: \\"should have required property '\\" + \\"id\\" + \\"'\\",
        schema: schema6.required,
        parentSchema: schema6,
        data,
      };

      if (vErrors === null) {
        vErrors = [err0];
      } else {
        vErrors.push(err0);
      }

      errors++;
    }

    if (data.name === undefined) {
      const err1 = {
        keyword: \\"required\\",
        dataPath,
        schemaPath: \\"#/required\\",
        params: {
          missingProperty: \\"name\\",
        },
        message: \\"should have required property '\\" + \\"name\\" + \\"'\\",
        schema: schema6.required,
        parentSchema: schema6,
        data,
      };

      if (vErrors === null) {
        vErrors = [err1];
      } else {
        vErrors.push(err1);
      }

      errors++;
    }

    if (data.id !== undefined) {
      let data0 = data.id;

      if (typeof data0 !== \\"string\\") {
        const err2 = {
          keyword: \\"type\\",
          dataPath: dataPath + \\"/id\\",
          schemaPath: \\"#/properties/id/type\\",
          params: {
            type: \\"string\\",
          },
          message: \\"should be string\\",
          schema: schema6.properties.id.type,
          parentSchema: schema6.properties.id,
          data: data0,
        };

        if (vErrors === null) {
          vErrors = [err2];
        } else {
          vErrors.push(err2);
        }

        errors++;
      }
    }

    if (data.name !== undefined) {
      let data1 = data.name;

      if (typeof data1 !== \\"string\\") {
        const err3 = {
          keyword: \\"type\\",
          dataPath: dataPath + \\"/name\\",
          schemaPath: \\"#/properties/name/type\\",
          params: {
            type: \\"string\\",
          },
          message: \\"should be string\\",
          schema: schema6.properties.name.type,
          parentSchema: schema6.properties.name,
          data: data1,
        };

        if (vErrors === null) {
          vErrors = [err3];
        } else {
          vErrors.push(err3);
        }

        errors++;
      }
    }
  } else {
    const err4 = {
      keyword: \\"type\\",
      dataPath,
      schemaPath: \\"#/type\\",
      params: {
        type: \\"object\\",
      },
      message: \\"should be object\\",
      schema: schema6.type,
      parentSchema: schema6,
      data,
    };

    if (vErrors === null) {
      vErrors = [err4];
    } else {
      vErrors.push(err4);
    }

    errors++;
  }

  validate21.errors = vErrors;
  return errors === 0;
}

const BlogPost = validate22;
const schema7 = {
  $id: \\"file:///BlogPost.json\\",
  title: \\"A Blog Post\\",
  description: \\"A blog post represents an article associated with an author\\",
  type: \\"object\\",
  properties: {
    id: {
      type: \\"string\\",
    },
    title: {
      type: \\"string\\",
    },
    content: {
      type: \\"string\\",
    },
    author: {
      $ref: \\"file:///User.json\\",
    },
  },
  required: [\\"id\\", \\"title\\", \\"content\\", \\"user\\"],
};

function validate22(
  data,
  { dataPath = \\"\\", parentData, parentDataProperty, rootData = data } = {}
) {
  let vErrors = null;
  let errors = 0;

  if (data && typeof data == \\"object\\" && !Array.isArray(data)) {
    if (data.id === undefined) {
      const err0 = {
        keyword: \\"required\\",
        dataPath,
        schemaPath: \\"#/required\\",
        params: {
          missingProperty: \\"id\\",
        },
        message: \\"should have required property '\\" + \\"id\\" + \\"'\\",
        schema: schema7.required,
        parentSchema: schema7,
        data,
      };

      if (vErrors === null) {
        vErrors = [err0];
      } else {
        vErrors.push(err0);
      }

      errors++;
    }

    if (data.title === undefined) {
      const err1 = {
        keyword: \\"required\\",
        dataPath,
        schemaPath: \\"#/required\\",
        params: {
          missingProperty: \\"title\\",
        },
        message: \\"should have required property '\\" + \\"title\\" + \\"'\\",
        schema: schema7.required,
        parentSchema: schema7,
        data,
      };

      if (vErrors === null) {
        vErrors = [err1];
      } else {
        vErrors.push(err1);
      }

      errors++;
    }

    if (data.content === undefined) {
      const err2 = {
        keyword: \\"required\\",
        dataPath,
        schemaPath: \\"#/required\\",
        params: {
          missingProperty: \\"content\\",
        },
        message: \\"should have required property '\\" + \\"content\\" + \\"'\\",
        schema: schema7.required,
        parentSchema: schema7,
        data,
      };

      if (vErrors === null) {
        vErrors = [err2];
      } else {
        vErrors.push(err2);
      }

      errors++;
    }

    if (data.user === undefined) {
      const err3 = {
        keyword: \\"required\\",
        dataPath,
        schemaPath: \\"#/required\\",
        params: {
          missingProperty: \\"user\\",
        },
        message: \\"should have required property '\\" + \\"user\\" + \\"'\\",
        schema: schema7.required,
        parentSchema: schema7,
        data,
      };

      if (vErrors === null) {
        vErrors = [err3];
      } else {
        vErrors.push(err3);
      }

      errors++;
    }

    if (data.id !== undefined) {
      let data0 = data.id;

      if (typeof data0 !== \\"string\\") {
        const err4 = {
          keyword: \\"type\\",
          dataPath: dataPath + \\"/id\\",
          schemaPath: \\"#/properties/id/type\\",
          params: {
            type: \\"string\\",
          },
          message: \\"should be string\\",
          schema: schema7.properties.id.type,
          parentSchema: schema7.properties.id,
          data: data0,
        };

        if (vErrors === null) {
          vErrors = [err4];
        } else {
          vErrors.push(err4);
        }

        errors++;
      }
    }

    if (data.title !== undefined) {
      let data1 = data.title;

      if (typeof data1 !== \\"string\\") {
        const err5 = {
          keyword: \\"type\\",
          dataPath: dataPath + \\"/title\\",
          schemaPath: \\"#/properties/title/type\\",
          params: {
            type: \\"string\\",
          },
          message: \\"should be string\\",
          schema: schema7.properties.title.type,
          parentSchema: schema7.properties.title,
          data: data1,
        };

        if (vErrors === null) {
          vErrors = [err5];
        } else {
          vErrors.push(err5);
        }

        errors++;
      }
    }

    if (data.content !== undefined) {
      let data2 = data.content;

      if (typeof data2 !== \\"string\\") {
        const err6 = {
          keyword: \\"type\\",
          dataPath: dataPath + \\"/content\\",
          schemaPath: \\"#/properties/content/type\\",
          params: {
            type: \\"string\\",
          },
          message: \\"should be string\\",
          schema: schema7.properties.content.type,
          parentSchema: schema7.properties.content,
          data: data2,
        };

        if (vErrors === null) {
          vErrors = [err6];
        } else {
          vErrors.push(err6);
        }

        errors++;
      }
    }

    if (data.author !== undefined) {
      let data3 = data.author;

      if (
        !validate21(data3, {
          dataPath: dataPath + \\"/author\\",
          parentData: data,
          parentDataProperty: \\"author\\",
          rootData,
        })
      ) {
        vErrors =
          vErrors === null
            ? validate21.errors
            : vErrors.concat(validate21.errors);
        errors = vErrors.length;
      }
    }
  } else {
    const err7 = {
      keyword: \\"type\\",
      dataPath,
      schemaPath: \\"#/type\\",
      params: {
        type: \\"object\\",
      },
      message: \\"should be object\\",
      schema: schema7.type,
      parentSchema: schema7,
      data,
    };

    if (vErrors === null) {
      vErrors = [err7];
    } else {
      vErrors.push(err7);
    }

    errors++;
  }

  validate22.errors = vErrors;
  return errors === 0;
}

const Codecs = {
  User: new CodecImpl(\\"User\\", \\"file:///User.json\\", User),
  BlogPost: new CodecImpl(\\"BlogPost\\", \\"file:///BlogPost.json\\", BlogPost),
};

exports.Codecs = Codecs;
exports.Types = types;
"
`;

exports[`Codec generation will produce the expected code 2`] = `
Object {
  "file:///BlogPost.json": "BlogPost",
  "file:///User.json": "User",
}
`;

exports[`Codec generation will produce the expected code 3`] = `
"interface Codec<T> {
    /**
     * Identify function returning the given argument as a value matching the schema.
     *
     * This can be useful to use in non-TypeScript code to construct a valid object while
     * benefitting from suggestions from a TypeScript language service.
     */
    identity(obj: T): T;
    /**
     * Check if a value matches the schema.
     */
    is(obj: unknown): obj is T;
    /**
     * Validate that a value matches the schema and throws if not.
     */
    validate(obj: unknown): T;
}
declare namespace Types {
    /**
     * A User Object
     *
     * A user is a known visitor.
     * @see file:///User.json
     */
    type User = {
        id: string;
        name: string;
    };
    /**
     * A Blog Post
     *
     * A blog post represents an article associated with an author
     * @see file:///BlogPost.json
     */
    type BlogPost = {
        id: string;
        title: string;
        content: string;
        author?: User;
    };
}
declare const Codecs: {
    readonly User: Codec<Types.User>;
    readonly BlogPost: Codec<Types.BlogPost>;
};
export { Codec, Types, Codecs };
"
`;
